File: go/nanoid/nanoid.go
// Package nanoid generates compact, URL-friendly random string IDs following
// https://github.com/ai/nanoid. Custom alphabets use mask-based rejection
// sampling so every symbol is equally likely.
package nanoid

import (
	"crypto/rand"
	"errors"
	"io"
	"math/bits"
	"unicode/utf8"
)

const (
	// URLAlphabet is the default 64-symbol URL-safe alphabet (A-Za-z0-9_-).
	URLAlphabet = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict"
	// DefaultSize gives a collision probability similar to UUID v4.
	DefaultSize = 21
)

var (
	ErrAlphabet = errors.New("nanoid: alphabet must contain between 1 and 256 symbols")
	ErrSize     = errors.New("nanoid: size must be >= 1")
)

// New returns a DefaultSize ID using URLAlphabet.
func New() string {
	var b [DefaultSize]byte
	fill(rand.Reader, b[:])
	for i := range b {
		b[i] = URLAlphabet[b[i]&63]
	}
	return string(b[:])
}

// NewSize returns an ID of the given size using URLAlphabet.
func NewSize(size int) (string, error) {
	if size < 1 {
		return "", ErrSize
	}
	b := make([]byte, size)
	fill(rand.Reader, b)
	for i := range b {
		b[i] = URLAlphabet[b[i]&63]
	}
	return string(b), nil
}

// Generator produces IDs from a fixed alphabet. Safe for concurrent use when
// its random source is (crypto/rand.Reader is).
type Generator struct {
	ascii  []byte
	runes  []rune
	length int
	mask   byte
	size   int
	random io.Reader
}

// CustomAlphabet returns a generator drawing from crypto/rand.
func CustomAlphabet(alphabet string, size int) (*Generator, error) {
	return CustomRandom(alphabet, size, rand.Reader)
}

// CustomRandom returns a generator using the supplied random byte source.
func CustomRandom(alphabet string, size int, random io.Reader) (*Generator, error) {
	if size < 1 {
		return nil, ErrSize
	}
	g := &Generator{size: size, random: random}
	if isASCII(alphabet) {
		g.ascii = []byte(alphabet)
		g.length = len(g.ascii)
	} else {
		g.runes = []rune(alphabet)
		g.length = len(g.runes)
	}
	if g.length < 1 || g.length > 256 {
		return nil, ErrAlphabet
	}
	g.mask = byte(2<<(bits.Len32(uint32(g.length-1)|1)-1) - 1)
	return g, nil
}

// Generate returns an ID of the generator's default size.
func (g *Generator) Generate() string { return g.generate(g.size) }

// GenerateSize returns an ID of the given size.
func (g *Generator) GenerateSize(size int) (string, error) {
	if size < 1 {
		return "", ErrSize
	}
	return g.generate(size), nil
}

func (g *Generator) step(size int) int {
	if int(g.mask)+1 == g.length {
		return size
	}
	m := int(g.mask)
	return (8*m*size + 5*g.length - 1) / (5 * g.length)
}

func (g *Generator) generate(size int) string {
	buf := make([]byte, g.step(size))
	if g.ascii != nil {
		out := make([]byte, 0, size)
		for {
			fill(g.random, buf)
			for _, b := range buf {
				if i := int(b & g.mask); i < g.length {
					out = append(out, g.ascii[i])
					if len(out) == size {
						return string(out)
					}
				}
			}
		}
	}
	out := make([]rune, 0, size)
	for {
		fill(g.random, buf)
		for _, b := range buf {
			if i := int(b & g.mask); i < g.length {
				out = append(out, g.runes[i])
				if len(out) == size {
					return string(out)
				}
			}
		}
	}
}

func fill(r io.Reader, b []byte) {
	if _, err := io.ReadFull(r, b); err != nil {
		panic("nanoid: random source failed: " + err.Error())
	}
}

func isASCII(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] >= utf8.RuneSelf {
			return false
		}
	}
	return true
}

File: rust/src/nanoid.rs
//! Nano ID: compact, URL-friendly random string IDs following
//! <https://github.com/ai/nanoid>. Custom alphabets use mask-based rejection
//! sampling so every symbol is equally likely.

use crate::{rng, Error};

/// Default 64-symbol URL-safe alphabet (`A-Za-z0-9_-`).
pub const URL_ALPHABET: &str = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict";
/// Default size, giving a collision probability similar to UUID v4.
pub const DEFAULT_SIZE: usize = 21;

/// A default-size ID using [`URL_ALPHABET`].
pub fn nanoid() -> String {
    let mut buf = [0u8; DEFAULT_SIZE];
    rng::fill(&mut buf);
    url_encode(&mut buf);
    // SAFETY: url_encode only writes ASCII bytes.
    unsafe { String::from_utf8_unchecked(buf.to_vec()) }
}

/// An ID of `size` symbols using [`URL_ALPHABET`].
pub fn nanoid_with_size(size: usize) -> Result<String, Error> {
    if size == 0 {
        return Err(Error::Size);
    }
    let mut buf = vec![0u8; size];
    rng::fill(&mut buf);
    url_encode(&mut buf);
    // SAFETY: url_encode only writes ASCII bytes.
    Ok(unsafe { String::from_utf8_unchecked(buf) })
}

fn url_encode(buf: &mut [u8]) {
    let alphabet = URL_ALPHABET.as_bytes();
    for b in buf.iter_mut() {
        *b = alphabet[(*b & 63) as usize];
    }
}

/// A generator bound to a custom alphabet and default size.
#[derive(Debug, Clone)]
pub struct CustomAlphabet {
    symbols: Vec<char>,
    ascii: bool,
    mask: u8,
    size: usize,
}

impl CustomAlphabet {
    pub fn new(alphabet: &str, size: usize) -> Result<Self, Error> {
        let symbols: Vec<char> = alphabet.chars().collect();
        let len = symbols.len();
        if !(1..=256).contains(&len) {
            return Err(Error::Alphabet);
        }
        if size == 0 {
            return Err(Error::Size);
        }
        let bits = u32::BITS - ((len as u32 - 1) | 1).leading_zeros();
        let mask = ((2u32 << (bits - 1)) - 1) as u8;
        Ok(CustomAlphabet { ascii: alphabet.is_ascii(), symbols, mask, size })
    }

    /// Generate an ID of the default size from the OS CSPRNG.
    pub fn generate(&self) -> String {
        self.generate_with(self.size, rng::fill)
    }

    /// Generate an ID of `size` symbols from the OS CSPRNG.
    pub fn generate_size(&self, size: usize) -> Result<String, Error> {
        if size == 0 {
            return Err(Error::Size);
        }
        Ok(self.generate_with(size, rng::fill))
    }

    /// Generate using a caller-supplied random source (nanoid's `customRandom`).
    pub fn generate_with<F: FnMut(&mut [u8])>(&self, size: usize, mut random: F) -> String {
        let len = self.symbols.len();
        let mask = self.mask as usize;
        let step = if mask + 1 == len { size } else { (8 * mask * size).div_ceil(5 * len) };
        let mut buf = vec![0u8; step];
        let mut out = String::with_capacity(if self.ascii { size } else { size * 4 });
        let mut count = 0;
        loop {
            random(&mut buf);
            for &b in &buf {
                let i = b as usize & mask;
                if i < len {
                    out.push(self.symbols[i]);
                    count += 1;
                    if count == size {
                        return out;
                    }
                }
            }
        }
    }
}
